src/VisualStudio/Core/Def/Implementation/Remote/GlobalNotificationRemoteDeliveryService.cs

```diff
@@ -0,0 +1,146 @@
+        private enum GlobalNotificationState
+        {
+            NotStarted,
+            Started,
+            Finished
+        }
+
+        /// <summary>
+        /// Lock for the <see cref="_globalNotificationsTask"/> task chain.  Each time we hear 
+        /// about a global operation starting or stopping (i.e. a build) we will '.ContinueWith'
+        /// this task chain with a new notification to the OOP side.  This way all the messages
+        /// are properly serialized and appera in the right order (i.e. we don't hear about a 
+        /// stop prior to hearing about the relevant start).
+        /// </summary>
+        private readonly object _globalNotificationsGate = new object();
+        private Task<GlobalNotificationState> _globalNotificationsTask = Task.FromResult(GlobalNotificationState.NotStarted);
+
+        private readonly HostWorkspaceServices _services;
+        private readonly CancellationToken _cancellationToken;
+
+        public GlobalNotificationRemoteDeliveryService(HostWorkspaceServices services, CancellationToken cancellationToken)
+        {
+            _services = services;
+            _cancellationToken = cancellationToken;
+
+            RegisterGlobalOperationNotifications();
+        }
+
+        public void Dispose()
+        {
+            UnregisterGlobalOperationNotifications();
+        }
+
+        private void RegisterGlobalOperationNotifications()
+        {
+            var globalOperationService = _services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started += OnGlobalOperationStarted;
+                globalOperationService.Stopped += OnGlobalOperationStopped;
+            }
+        }
+
+        private void UnregisterGlobalOperationNotifications()
+        {
+            var globalOperationService = _services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started -= OnGlobalOperationStarted;
+                globalOperationService.Stopped -= OnGlobalOperationStopped;
+            }
+        }
+
+        private void OnGlobalOperationStarted(object sender, EventArgs e)
+        {
+            lock (_globalNotificationsGate)
+            {
+                _globalNotificationsTask = _globalNotificationsTask.SafeContinueWithFromAsync(
+                    continuation, _cancellationToken, TaskContinuationOptions.None, TaskScheduler.Default);
+            }
+
+            async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> previousTask)
+            {
+                // Can only transition from NotStarted->Started.  If we hear about
+                // anything else, do nothing.
+                if (previousTask.Result != GlobalNotificationState.NotStarted)
+                {
+                    return previousTask.Result;
+                }
```

Comment:
should we assert we're not in the 'Finished' state?  can we oc what hte Finished state actually is?

Reply:
n/m.  i read this wrong.